### PR TITLE
Debug HUD flag

### LINKED ISSUE #12011: Grant `basic_debug` privilege by default
(please add this to 5.5.0 milestone)

I am the one who made the PR that hides part of the debug screen by default (coordinates, yaw, pointed node, entity HP, seed, etc. but not FPS, version, lag, etc.). The rationale was that I wanted to allow games to stop players from getting this info for free. This makes sense for games when players must craft a compass first or whatever. See: #9315

I also added the `basic_debug` privilege which basically gives players those gameplay info back.

After having thought long about it (and many players having complained in the forums), I have come to the conclusion that maybe the change as it stands was not so great after all. My previous argument was that servers should just grant the `basic_debug` priv to get back the old behavior. But then I realized that some servers can be *notoriously* slow to react to such things. I'm afraid that the 5.5.0 release will de facto cause many players lose the ONLY convenient (although hacky) access to coordinates on servers on which it matters.

I'm afraid this can be a major disruption to players and annoy a lot of players. I heard they are already preparing the pitchforks … :D Releasing 5.5.0 with this change as it stands now might be bad.

So I have an idea on how to avoid this disruption in a simple way: Grant the `basic_debug` privilege by default. That way, the games that want to hide the info can now hide it and longstanding server playerbases will NOT be affected.

This change sounds very simple and easy, right? Well, not really. Because privileges granted by default to singleplayer cannot be revoked. But if this privilege cannot be revoked, it is pretty pointless. :D
I don't know if this is a bug but either way, this means that before `basic_debug` can be granted by default, it needs to be possible to revoke default privileges. I don't know if this counts as 'feature'.

Either way, I think this issue must be dealt somehow before 5.5 release. Having hundreds of players lose access to coordinates after an upgrade is too high a price to pay. I previously would just have blamed server operators for that but this is an unrealistic expectation.

If this can not be done in time before release, I sadly see no other option than a full revert of #9315 for now. :-(

### PR DESCRIPTION
**Pull request by appgurueu**
Alternative to #12016: Closes #12011 by converting `basic_debug` into a HUD flag which, like all HUD flags, is `true` by default.

A couple notes:

* ~~This includes removing the strings from translations, which is the reason for the large diff~~
* The flag must not be sent inverted or anything due to the server-sent mask which should never have the debug bit set on older servers
* Minor cleanup of the related code included

Ready for review.

### CODE REVIEW

**Comment by LoneWolfHT** [comment]
For testing:

```lua
function give_debug()
	for _, player in ipairs(minetest.get_connected_players()) do
		player:hud_set_flags({
			debug = true,
		})
	end

	return true, "Gave debug"
end

function take_debug()
	for _, player in ipairs(minetest.get_connected_players()) do
		player:hud_set_flags({
			debug = false,
		})
	end

	return true, "Took debug"
end

minetest.register_chatcommand("give_debug", {
	func = give_debug
})

minetest.register_chatcommand("take_debug", {
	func = take_debug
})
```

[basic_debug.zip](https://github.com/minetest/minetest/files/7966867/basic_debug.zip)

**Comment by LoneWolfHT** [comment]
When you join a world the hotbar isn't shown by default

**Review by LoneWolfHT — APPROVED** [review]
Tested

* Other HUD flags still seem to work
* debug hud flag toggling works as expected
* New clients on old servers have basic_debug enabled
* New clients on new servers have basic_debug enabled by default

Code looks good

**Comment by lhofhansl** [comment]
Looks good in principle and I support this way of debug info.
(Haven't had time to test, though)

**Comment by Wuzzy2** [comment]
This needs extensive testing, I don't think it will make it in time for 5.5.0, we're in feature freeze anyway.

**Comment by appgurueu** [comment]
> This needs extensive testing, I don't think it will make it in time for 5.5.0, we're in feature freeze anyway.

This isn't a new feature in my book though - it just swaps out / changes the implementation between the feature you already introduced to (1) use HUD flags and (2) (through this) default to true.

**Comment by appgurueu** [comment]
Rebased.

File: po/de/minetest.po [hunk 1]
```diff
@@ -1278,7 +1278,7 @@ msgid "Camera update enabled"
 msgstr "Kameraaktualisierung aktiviert"
 
 #: src/client/game.cpp
-msgid "Can't show block bounds (need 'basic_debug' privilege)"
+msgid "Can't show block bounds (disabled by mod or game)"
 msgstr ""
 "Blockgrenzen können nicht gezeigt werden („basic_debug“-Privileg benötigt)"
```
Comment (Df458) [inline po/de/minetest.po:1283]: Line remains

File: po/fr/minetest.po [hunk 2]
```diff
@@ -1279,7 +1279,7 @@ msgid "Camera update enabled"
 msgstr "Mise à jour de la caméra activée"
 
 #: src/client/game.cpp
-msgid "Can't show block bounds (need 'basic_debug' privilege)"
+msgid "Can't show block bounds (disabled by mod or game)"
 msgstr ""
 "Impossible d'afficher les limites des blocs (nécessite le privilège "
 "« basic_debug »)"
```
Comment (Df458) [inline po/fr/minetest.po:1285]: Line remains

File: po/hu/minetest.po [hunk 3]
```diff
@@ -1276,7 +1276,7 @@ msgid "Camera update enabled"
 msgstr "Kamera frissítés engedélyezve"
 
 #: src/client/game.cpp
-msgid "Can't show block bounds (need 'basic_debug' privilege)"
+msgid "Can't show block bounds (disabled by mod or game)"
 msgstr ""
 "Blokkhatárok mutatása nem lehetséges ('basic_debug' jogosultság szükséges)"
```
Comment (Df458) [inline po/hu/minetest.po:1281]: Line remains

File: po/it/minetest.po [hunk 4]
```diff
@@ -1276,7 +1276,7 @@ msgid "Camera update enabled"
 msgstr "Aggiornamento telecamera abilitato"
 
 #: src/client/game.cpp
-msgid "Can't show block bounds (need 'basic_debug' privilege)"
+msgid "Can't show block bounds (disabled by mod or game)"
 msgstr ""
 "Impossibile mostrare i limiti del blocco (si ha bisogno del privilegio "
 "'baisc_debug')"
```
Comment (Df458) [inline po/it/minetest.po:1282]: Line remains

**Review by Df458 — CHANGES_REQUESTED** [review]
Code looks good at first glance, but I haven't tested yet. A few localization lines haven't been cleared yet, I've marked them for you.

File: po/de/minetest.po (hunk 1 above)
Comment (rubenwardy) [reply po/de/minetest.po:1283]: These files are updated by the translation update script, and shouldn't be modified in normal PRs

**Comment by appgurueu** [comment]
> Code looks good at first glance, but I haven't tested yet. A few localization lines haven't been cleared yet, I've marked them for you.

Ah indeed, fancy quotes mean my grepping failed.

**Comment by rubenwardy** [comment]
Please don't

File: po/ar/minetest.po [hunk 5]
```diff
@@ -1277,8 +1277,8 @@ msgid "Camera update enabled"
 msgstr "تحديث الكاميرا مفعل"
 
 #: src/client/game.cpp
-msgid "Can't show block bounds (need 'basic_debug' privilege)"
-msgstr "تعذر إظهار حدود الكتل ( تحتاج امتياز 'basic_debug')"
+msgid "Can't show block bounds (disabled by mod or game)"
+msgstr ""
```
Comment (rubenwardy) [inline po/ar/minetest.po:1281]: Please don't change .po files in files, as it conflicts with translation updates

**Comment by appgurueu** [comment]
> Please don't

So should I just remove the translation-related removals?

File: doc/lua_api.txt [hunk 6]
```diff
@@ -6753,17 +6753,16 @@ object you are working with still exists.
...
-        * minimap_radar
+        * `hotbar`
+        * `healthbar`
+        * `crosshair`
+        * `wielditem`
+        * `breathbar`
+        * `minimap`: Modifies the client's permission to view the minimap.
+          The client may locally elect to not view the minimap.
+        * `minimap_radar`: is only usable when `minimap` is true
+        * `debug`: Allow showing basic debug info (pos etc.)
```
Comment (rubenwardy) [inline doc/lua_api.txt:6764]: This might be confusing given that the debug priv and wireframe feature exists

File: src/client/game.cpp [hunk 7]
```diff
@@ -2298,18 +2296,19 @@ void Game::toggleFog()
 
 void Game::toggleDebug()
 {
+	LocalPlayer *player = client->getEnv().getLocalPlayer();
+	bool has_basic_debug = player->hud_flags & HUD_FLAG_DEBUG;
 	// Initial: No debug info
 	// 1x toggle: Debug text
 	// 2x toggle: Debug text with profiler graph
-	// 3x toggle: Debug text and wireframe (needs "debug" priv)
+	// 3x toggle: Debug text and wireframe
```
Comment (rubenwardy) [inline src/client/game.cpp:2304]: Still needs debug priv

**Comment by rubenwardy** [comment]
> > Please don't
> 
> So should I just remove the translation-related removals?

Yeah, please remove all changes to .po files

(`git checkout upstream/master -- locale` is how I tend to do that)

File: src/client/game.cpp (hunk 7 above)
Comment (appgurueu) [reply src/client/game.cpp:2304]: Yeah, must've left that change in when ripping out the debug priv removal...

File: doc/lua_api.txt (hunk 6 above)
Comment (appgurueu) [reply doc/lua_api.txt:6764]: I'll just take @Wuzzy2's description of the basic_debug priv instead

**Review by Df458 — APPROVED** [review]

**Comment by Wuzzy2** [comment]
Damn. So it *does* seem the HUD flag approach is superior after all.

So this PR looks like it's exactly the same as my priv-based PR, but with HUD flags instead. It hides exactly the HUD elements it should hide. It reacts correctly with regards to entity infotest or block bounds.

And it has zero of the drawbacks that came from using privs. With this change, it doesn't look like old players on servers won't suddenly lose their coords after the server upgraded (unless the admin intervenes. This is an important difference that the priv-based "solution" failed in. Your PR doesn't seem to require any admin intervention; after the update to the next version, everything should stay the same unless there's explicit admin intervention.
And since the HUD flag defaults to true, this should (hopefully) be backwards-compatible on servers, unless I overlook something.

I'm glad I decided to push for a revert on my priv-based approach.

The reason why I was against HUD flags initially was purely sentimental, I have to admit now. It just "felt weird" to me that a *debug* screen (something technical) is controlled by a HUD flag (something something gameplay). This obviously is a bad argument.

One thing I did not test was connecting with old clients. What happens if an old client connects to a new server?

**Comment by appgurueu** [comment]
> One thing I did not test was connecting with old clients. What happens if an old client connects to a new server?

It should simply ignore the debug HUD flag and always show the debug HUD.

**Comment by x2048** [comment]
May I suggest two things here:
* Keep the map seed out of 'debug' HUD flag, because it does not affect gameplay, and is not a debug-only information.
* Still show the debug information in "-dev" builds (or have an easy way to enable it). It will be unfortunate if it becomes complicated for play-testers to share info with the engine developers.

**Comment by appgurueu** [comment]
> Keep the map seed out of 'debug' HUD flag, because it does not affect gameplay, and is not a debug-only information.

Out of scope for this PR.

> Still show the debug information in "-dev" builds (or have an easy way to enable it). It will be unfortunate if it becomes complicated for play-testers to share info with the engine developers.

We don't want any "cheats" based on whether a version is a development version. If a mod explicitly sets the flag to `false` for a player, the player is not meant to see the debug information, no matter their build. Mods will have to define which players they consider "playtesters" or "debuggers" who they allow access to the HUD. A simple solution might be just allowing those with the `debug` priv access, but as said, that's up to modders to decide - this PR merely gives them an API to take control of it.

Besides, what useful information does the debug HUD display anyways? Despite it's name, I have hardly seen it being used for debugging; it's primary purpose up to this point was to provide players a means to display their coords and to know what they are looking at (albeit in technical, user-unfriendly terms).

**Comment by Wuzzy2** [comment]
Oh, right. This is a good idea by appgurueu.
If the player has the `debug` priv, the player is *always* able to see the full debug screen, thus bypassing the HUD flag. This allows admins to continue working under any game.
It makes complete sense.

**Comment by appgurueu** [comment]
> Oh, right. This is a good idea by appgurueu. If the player has the `debug` priv, the player is _always_ able to see the full debug screen, thus bypassing the HUD flag. This allows admins to continue working under any game. It makes complete sense.

Should I add that to this PR, thereby making this engine-defined and reducing flexibility, or should we just leave this as a recommendation for mods controlling the HUD flag?

**Comment by Wuzzy2** [comment]
Yes, please add this.

File: doc/lua_api.txt [hunk 8]
```diff
@@ -6753,17 +6753,18 @@ object you are working with still exists.
...
-        * minimap_radar
+        * `hotbar`
+        * `healthbar`
+        * `crosshair`
+        * `wielditem`
+        * `breathbar`
+        * `minimap`: Modifies the client's permission to view the minimap.
+          The client may locally elect to not view the minimap.
+        * `minimap_radar`: is only usable when `minimap` is true
+        * `debug`: Allow showing basic debug info that might give a gameplay advantage.
```
Comment (sfan5) [inline doc/lua_api.txt:6764]: My primary and obvious complain is here: shouldn't this be renamed `basic_debug` so people do not confuse it with the `debug` privilege?

Comment (appgurueu) [reply doc/lua_api.txt:6764]: Now that the debug priv does indeed override this, probably yes.

Comment (appgurueu) [reply doc/lua_api.txt]: Last commit does this.

**Review by Df458 — COMMENTED** [review]
...dare I ask why there appear to be a ton of unrelated particlesystem changes in this PR now?

**Comment by appgurueu** [comment]
> ...dare I ask why there appear to be a ton of unrelated particlesystem changes in this PR now?

Git messup: I had tested the particle PR and must've forgotten to switch branches at some point. Should be fixed now (resetted to before the merge commit and cherry-picked the HUD flag rename).

**Review by Df458 — APPROVED** [review]
Gave this a brief pass just to verify that it still works and uses `basic_debug`. Seems fine now.

File: doc/lua_api.txt [hunk 9]
```diff
@@ -6753,17 +6753,18 @@ object you are working with still exists.
...
+        * `healthbar`
+        * `crosshair`
+        * `wielditem`
+        * `breathbar`
+        * `minimap`: Modifies the client's permission to view the minimap.
+          The client may locally elect to not view the minimap.
+        * `minimap_radar`: is only usable when `minimap` is true
+        * `basic_debug`: Allow showing basic debug info that might give a gameplay advantage.
+          This includes map seed, player position, look direction, the pointed node and block bounds.
+		  Does not affect players with the `debug` privilege.
```
Comment (Desour) [inline doc/lua_api.txt:6766]: Tabs detected.

Comment (appgurueu) [reply doc/lua_api.txt]: Fixed.

**Comment by sfan5** [comment]
`//lua core.get_player_by_name("singleplayer"):hud_set_flags({basic_debug=false})`
and nothing at all happens, what am I doing wrong?

**Comment by sfan5** [comment]
That was it, thanks.

**Review by sfan5 — APPROVED** [review]

**Pull request merged**